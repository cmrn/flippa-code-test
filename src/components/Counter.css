.Counter {
  padding: 10px 0;
  display: flex;
  border-bottom: 1px solid #ccc;
  align-items: center;
  font-size: 20px;
}

.Counter:first-of-type {
  border-top: 1px solid #ccc;
}

.Counter-title {
  flex-grow: 1;
  padding: 0 10px;
}

.Counter-count {
  padding: 0 10px;
}

.Counter-button {
  border: none;
  background: none;
}

.Counter-delete {
  font-size: 15px;
}