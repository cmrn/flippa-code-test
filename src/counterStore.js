import { store } from 'react-easy-state';
import { 
  addCounter, 
  getCounters, 
  incrementCounter, 
  decrementCounter, 
  deleteCounter
} from './api';

const counterStore = store({
  counters: {},
  async load() {
    const response = await getCounters();
    handleApiResponse(response);
  },
  reset() {
    counterStore.counters = {};
  },
  async add(title) {
    const response = await addCounter(title);
    handleApiResponse(response);
  },
  async increment(id) {
    const counter = counterStore.get(id);
    counter.count++;
    const response = await incrementCounter(id);
    handleApiResponse(response);
  },
  async decrement(id) {
    const counter = counterStore.get(id);
    counter.count--;
    const response = await decrementCounter(id);
    handleApiResponse(response);
  },
  async delete(id) {
    const response = await deleteCounter(id);
    handleApiResponse(response);
  },
  get(id) {
    const counter = counterStore.counters[id];
    if(!counter) throw new Error(`invalid counter id ${id}`);
    return counter;
  }
});

function handleApiResponse(countersArray) {
  const countersObj = countersArray.reduce((acc, counter) => {
    acc[counter.id] = counter;
    return acc;
  }, {});
  counterStore.counters = countersObj;
};

export default counterStore;
