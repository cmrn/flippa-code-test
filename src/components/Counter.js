import React from 'react';
import PropTypes from 'prop-types';
import './Counter.css';
import FontAwesomeIcon from '@fortawesome/react-fontawesome';
import faMinus from '@fortawesome/fontawesome-free-solid/faMinus';
import faPlus from '@fortawesome/fontawesome-free-solid/faPlus';
import faTimes from '@fortawesome/fontawesome-free-solid/faTimes';

const Counter = ({ title, count, onChange }) => (
  <div className="Counter">
    <button className="Counter-button Counter-delete">
      <FontAwesomeIcon color="crimson" icon={faTimes} />
    </button>
    <span className="Counter-title">{title}</span>
    <button className="Counter-button Counter-subtract" onClick={() => onChange(count-1)}>
      <FontAwesomeIcon icon={faMinus} />
    </button>
    <span className="Counter-count">{count}</span>
    <button className="Counter-button Counter-add" onClick={() => onChange(count+1)}>
      <FontAwesomeIcon icon={faPlus} />
    </button>
  </div>
);

Counter.propTypes = {
  title: PropTypes.string.isRequired,
  count: PropTypes.number.isRequired,
  onChange: PropTypes.func.isRequired,
};

export default Counter;
