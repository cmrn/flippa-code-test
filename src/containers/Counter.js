import React from 'react';
import Counter from '../components/Counter';
import { view } from 'react-easy-state';
import counterStore from '../counterStore';

const CounterContainer = ({ id, ...props }) => {
  const counter = counterStore.counters[id];
  if(!counter) return null;
  return (
    <Counter
      {...props}
      title={counter.title}
      count={counter.count}
      onIncrement={() => counterStore.increment(counter.id)}
      onDecrement={() => counterStore.decrement(counter.id)}
      onDelete={() => counterStore.delete(counter.id)}
    />
  );
};

export default view(CounterContainer);
