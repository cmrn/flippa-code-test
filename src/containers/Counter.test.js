import React from 'react';
import { shallow } from 'enzyme';
import Counter from './Counter';
import counterStore from '../counterStore';


const data = [
  {id: "asdf", title: "bob", count: 1},
  {id: "qwer", title: "steve", count: -1}
];

beforeEach(() => {
  counterStore.counters = data;
  counterStore.add = jest.fn();
  counterStore.increment = jest.fn();
  counterStore.decrement = jest.fn();
  counterStore.delete = jest.fn();
});

it('loads counter from counterStore by id', () => {
  const wrapper = shallow(<Counter id='qwer' />);
  expect(wrapper).toMatchSnapshot();
});

describe('when increment event is triggered', () => {
  it('calls counterStore.increment() with counter ID', () => {
    const id = 'qwer';
    const wrapper = shallow(<Counter id={id} />);
    expect(counterStore.increment.mock.calls.length).toBe(0);
    wrapper.props().onIncrement();
    expect(counterStore.increment.mock.calls.length).toBe(1);
    expect(counterStore.increment.mock.calls[0][0]).toBe(id);
  });
});

describe('when decrement event is triggered', () => {
  it('calls counterStore.decrement() with counter ID', () => {
    const id = 'qwer';
    const wrapper = shallow(<Counter id={id} />);
    expect(counterStore.decrement.mock.calls.length).toBe(0);
    wrapper.props().onDecrement();
    expect(counterStore.decrement.mock.calls.length).toBe(1);
    expect(counterStore.decrement.mock.calls[0][0]).toBe(id);
  });
});

describe('when delete event is triggered', () => {
  it('calls counterStore.delete() with counter ID', () => {
    const id = 'qwer';
    const wrapper = shallow(<Counter id={id} />);
    expect(counterStore.delete.mock.calls.length).toBe(0);
    wrapper.props().onDelete();
    expect(counterStore.delete.mock.calls.length).toBe(1);
    expect(counterStore.delete.mock.calls[0][0]).toBe(id);
  });
});
