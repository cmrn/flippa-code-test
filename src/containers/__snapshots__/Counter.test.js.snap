// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loads counter from counterStore by id 1`] = `
<Counter
  count={-1}
  onDecrement={[Function]}
  onDelete={[Function]}
  onIncrement={[Function]}
  title="steve"
/>
`;

exports[`renders nothing if counter id is invalid 1`] = `""`;
